### PR TITLE
SDI-266 New refactored create booking service

### PR DESCRIPTION
**Pull request by andymarke**
Existing very good integration tests unchanged showing now behaviour changes

* TODO Restricted patients calls old method, need to understand if that is still need before removing old version

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import lombok.RequiredArgsConstructor;
 import lombok.Setter;
 import lombok.ToString.Exclude;
+import lombok.With;
 import org.hibernate.Hibernate;
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java:12]: Kotlin can't see builder methods but can see `with` methods

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/AgencyInternalLocationRepositoryExt.kt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package uk.gov.justice.hmpps.prison.repository.jpa.repository
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/AgencyInternalLocationRepositoryExt.kt:1]: added kotlin extensions where we convert a Optional into a nullable similar to generic extension for `findById`

File: src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/ExternalMovementTransferService.kt [hunk 3]
```diff
@@ -160,12 +198,17 @@ class ExternalMovementTransferService(
...
+      ?: failure(EntityNotFoundException.withMessage("No movement reason $movementReasonCode found"))
+
+  fun getReceiveDateTime(movementTime: LocalDateTime?, booking: OffenderBooking?): Result<LocalDateTime> {
     val now = LocalDateTime.now()
     return movementTime?.let {
       return if (movementTime.isAfter(now)) {
         failure(BadRequestException("Transfer cannot be done in the future"))
-      } else if (booking.hasMovementsAfter(movementTime)) {
+      } else if (booking?.hasMovementsAfter(movementTime) == true) {
         failure(BadRequestException("Movement cannot be before the previous active movement"))
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/ExternalMovementTransferService.kt:212]: new offenders might not have a previous booking so allow nulls here

File: src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/ExternalMovementTransferService.kt [hunk 4]
```diff
@@ -139,6 +139,44 @@ class ExternalMovementTransferService(
...
+    commentText: String
+  ): ExternalMovement {
+    val movementReason = getMovementReasonForNewBooking(movementReasonCode).getOrThrow()
+    val movementType = getAdmissionMovementType().getOrThrow()
+
+    return with(booking) {
+      setPreviousMovementsToInactive().also { entityManager.flush() }
+      addExternalMovement(
+        ExternalMovement(
+          /* offenderBooking = */ booking,
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/ExternalMovementTransferService.kt:157]: Why commented out?

File: src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/BookingIntoPrisonService.kt [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+      agencyLocationRepository.findByIdAndTypeAndActiveAndDeactivationDateIsNullOrNull(prisonId, it, true)
+        ?.let { prison -> success(prison) } ?: failure(
+        EntityNotFoundException.withMessage("$prisonId prison not found")
+      )
+    } ?: failure(
+      EntityNotFoundException.withMessage("Not Found")
+    )
+
+  private fun cellOrReceptionCode(prisonId: String, cellLocation: String?) =
+    cellLocation?.takeIf { it.isNotBlank() } ?: "$prisonId-RECP"
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/BookingIntoPrisonService.kt:178]: How do we know every prison has this cell?

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**